From d48981ad7e36abb3500161d823acf92345c94f5d Mon Sep 17 00:00:00 2001
Message-Id: <d48981ad7e36abb3500161d823acf92345c94f5d.1571917458.git.pmatilai@redhat.com>
In-Reply-To: <68d383c39cef8d58b80940b13dd132d3f41a03f0.1571917458.git.pmatilai@redhat.com>
References: <68d383c39cef8d58b80940b13dd132d3f41a03f0.1571917458.git.pmatilai@redhat.com>
From: Panu Matilainen <pmatilai@redhat.com>
Date: Tue, 2 Apr 2019 16:07:56 +0300
Subject: [PATCH 2/2] Support build-id generation from compressed ELF files
 (elfutils >= 0.175)

Use dwelf_elf_begin() for reading ELF files for build-id generation on
versions that have it to support compressed ELF files such as kernel
modules (RhBug:1650072,1650074). Note that debugedit still cannot handle
compressed files, this is only for build-id generation.
---
 build/files.c | 4 ++++
 configure.ac  | 4 ++++
 2 files changed, 8 insertions(+)

diff --git a/build/files.c b/build/files.c
index 3822be3d3..f72a7c866 100644
--- a/build/files.c
+++ b/build/files.c
@@ -1828,7 +1828,11 @@ static int generateBuildIDs(FileList fl, ARGV_t *files)
 		   kernel modules (ET_REL files with .modinfo section)
 		   should have build-ids. */
 		GElf_Ehdr ehdr;
+#if HAVE_DWELF_ELF_BEGIN
+		Elf *elf = dwelf_elf_begin(fd);
+#else
 		Elf *elf = elf_begin (fd, ELF_C_READ, NULL);
+#endif
 		if (elf != NULL && elf_kind(elf) == ELF_K_ELF
 		    && gelf_getehdr(elf, &ehdr) != NULL
 		    && (ehdr.e_type == ET_EXEC || ehdr.e_type == ET_DYN
diff --git a/configure.ac b/configure.ac
index 99ce7df32..b2d7ed806 100644
--- a/configure.ac
+++ b/configure.ac
@@ -487,6 +487,10 @@ AS_IF([test "$WITH_LIBELF" = yes],[
       # If possible we also want the strtab functions from elfutils 0.167.
       # But we can fall back on the (unsupported) ebl alternatives if not.
       AC_CHECK_LIB(dw, dwelf_strtab_init, [HAVE_LIBDW_STRTAB=yes])
+      # whether libdw supports compressed ELF objects
+      AC_CHECK_LIB(dw, dwelf_elf_begin, [
+                   AC_DEFINE(HAVE_DWELF_ELF_BEGIN, 1, [Have dwelf_elf_begin?])
+      ])
     ])
   ])
 ])
-- 
2.21.0

